js/ui/scheduler/utils.recurrence.js

```diff
@@ -1,854 +1,413 @@
+                result = rule['byday'];
+            } else {
+                result = rule['byday'].split(',');
             }
         }
-        needCorrect && correctDate(date, expectedDate);
-    }
 
-    if(rule.freq === 'YEARLY') {
-        if(rule['byyearday']) {
-            const dayNumber = Number(rule['byyearday'].split(',')[iterationStep]);
-            dateSetterMap['byyearday'](date, dayNumber);
-        }
-
-        const dateRules = splitDateRules(rule);
-
-        for(const field in dateRules[iterationStep]) {
-            dateSetterMap[field] && dateSetterMap[field](date, dateRules[iterationStep][field], rule['wkst']);
-        }
-    }
-
-    return date;
-}
-
-function getDatePartDiffs(date1, date2) {
-    return {
-        years: date1.getFullYear() - date2.getFullYear(),
-        months: date1.getMonth() - date2.getMonth(),
-        days: date1.getDate() - date2.getDate(),
-        hours: date1.getHours() - date2.getHours(),
-        minutes: date1.getMinutes() - date2.getMinutes(),
-        seconds: date1.getSeconds() - date2.getSeconds()
-    };
-}
-
-function getRecurrenceRule(recurrence) {
-    const result = {
-        rule: {},
-        isValid: false
-    };
-
-    if(recurrence) {
-        result.rule = parseRecurrenceRule(recurrence);
-        result.isValid = validateRRule(result.rule, recurrence);
-    }
-
-    return result;
-}
-
-const loggedWarnings = [];
-
-function validateRRule(rule, recurrence) {
-    if(brokenRuleNameExists(rule) ||
-        inArray(rule.freq, freqNames) === -1 ||
-        wrongCountRule(rule) || wrongIntervalRule(rule) ||
-        wrongDayOfWeek(rule) ||
-        wrongByMonthDayRule(rule) || wrongByMonth(rule) ||
-        wrongUntilRule(rule)) {
-
-        logBrokenRule(recurrence);
-        return false;
-    }
-
-    return true;
-}
-
-function wrongUntilRule(rule) {
-    let wrongUntil = false;
-    const until = rule.until;
-
-    if(until !== undefined && !(until instanceof Date)) {
-        wrongUntil = true;
+        return result;
     }
 
-    return wrongUntil;
-}
-
-function wrongCountRule(rule) {
-    let wrongCount = false;
-    const count = rule.count;
+    getAsciiStringByDate(date) {
+        const currentOffset = this.getTimeZoneOffset() * toMs('minute');
 
-    if(count && typeof count === 'string') {
-        wrongCount = true;
+        date = new Date(date.getTime() + currentOffset);
+        return date.getFullYear() + ('0' + (date.getMonth() + 1)).slice(-2) + ('0' + date.getDate()).slice(-2) +
+            'T' + ('0' + (date.getHours())).slice(-2) + ('0' + (date.getMinutes())).slice(-2) + ('0' + (date.getSeconds())).slice(-2) + 'Z';
     }
 
-    return wrongCount;
-}
-
-function wrongByMonthDayRule(rule) {
-    let wrongByMonthDay = false;
-    const byMonthDay = rule['bymonthday'];
-
-    if(byMonthDay && isNaN(parseInt(byMonthDay))) {
-        wrongByMonthDay = true;
-    }
+    getRecurrenceString(object) {
+        if(!object || !object.freq) {
```

Comment:
if(!object?.freq) {

Reply:
I don't like this approach.
@andrewmakarov what do you think?

Reply:
optional chaining doesn'n support by ie11

Reply:
If I am not mistaken, we can use it since this [PR](https://github.com/DevExpress/DevExtreme/pull/11291). But, of course, I do not insist.